Script to alter the til tracking numpy arrays if an error has occurred

In [1]:
# Standard packages
import tempfile
import warnings
import urllib
import shutil
import os
# Less standard, but still pip- or conda-installable
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
#import rasterio
import re
import rtree
import shapely
import pickle
import data_eng.az_proc as ap
#from cartopy import crs
import collections

In [2]:
import PIL
print('PIL',PIL.__version__)
from PIL import Image
Image.MAX_IMAGE_PIXELS = 999999999999

PIL 8.3.1


Specify the parameters to download images

In [3]:
parent_directory = "C:\chips_allocation"
annotation_directory = "Check"
number_of_tiles =60
tiles_remaining = "tile_name_tile_url_remaining_expanded.npy"
tiles_labeled = "tile_name_tile_url_labeled.npy"

Load the annotator class

In [5]:
dist = ap.annotator(annotation_directory) #create the processing class
dist.state_dcc_directory(parent_directory)
dist.number_of_tiles(number_of_tiles)
dist.get_tile_urls(tiles_remaining)
dist.make_subdirectories()
dist.download_images()
#dist.tile_rename()

Unlabeled Tiles (571, 2)
0
Bypassing download of already-downloaded file m_2909636_nw_14_060_20181201_20190426.tif
1


100% (539177490 of 539177490) |##########| Elapsed Time: 0:01:20 Time:  0:01:20


...done, 539177490 bytes.
2


100% (613549436 of 613549436) |##########| Elapsed Time: 0:01:04 Time:  0:01:04


...done, 613549436 bytes.
3


100% (518720020 of 518720020) |##########| Elapsed Time: 0:01:39 Time:  0:01:39


...done, 518720020 bytes.
4


100% (592886951 of 592886951) |##########| Elapsed Time: 0:01:11 Time:  0:01:11


...done, 592886951 bytes.
5


100% (592466651 of 592466651) |##########| Elapsed Time: 0:01:21 Time:  0:01:21


...done, 592466651 bytes.
6


100% (472887819 of 472887819) |##########| Elapsed Time: 0:00:46 Time:  0:00:46


...done, 472887819 bytes.
7


100% (517730627 of 517730627) |##########| Elapsed Time: 0:00:33 Time:  0:00:33


...done, 517730627 bytes.
8


100% (598664739 of 598664739) |##########| Elapsed Time: 0:01:54 Time:  0:01:54


...done, 598664739 bytes.
9


100% (428623269 of 428623269) |##########| Elapsed Time: 0:00:16 Time:  0:00:16


...done, 428623269 bytes.
10


100% (621799797 of 621799797) |##########| Elapsed Time: 0:01:06 Time:  0:01:06


...done, 621799797 bytes.
11


100% (438033760 of 438033760) |##########| Elapsed Time: 0:01:06 Time:  0:01:06


...done, 438033760 bytes.
12


100% (475813115 of 475813115) |##########| Elapsed Time: 0:00:53 Time:  0:00:53


...done, 475813115 bytes.
13


100% (593778864 of 593778864) |##########| Elapsed Time: 0:00:42 Time:  0:00:42


...done, 593778864 bytes.
14


100% (633054698 of 633054698) |##########| Elapsed Time: 0:01:29 Time:  0:01:29


...done, 633054698 bytes.
15


100% (527398320 of 527398320) |##########| Elapsed Time: 0:01:21 Time:  0:01:21


...done, 527398320 bytes.
16


100% (518674697 of 518674697) |##########| Elapsed Time: 0:00:56 Time:  0:00:56


...done, 518674697 bytes.
17


100% (585514213 of 585514213) |##########| Elapsed Time: 0:00:48 Time:  0:00:48


...done, 585514213 bytes.
18


100% (559662736 of 559662736) |##########| Elapsed Time: 0:01:35 Time:  0:01:35


...done, 559662736 bytes.
19


100% (576420628 of 576420628) |##########| Elapsed Time: 0:00:36 Time:  0:00:36


...done, 576420628 bytes.
20


100% (587172670 of 587172670) |##########| Elapsed Time: 0:01:00 Time:  0:01:00


...done, 587172670 bytes.
21


100% (509213267 of 509213267) |##########| Elapsed Time: 0:01:27 Time:  0:01:27


...done, 509213267 bytes.
22


100% (519873399 of 519873399) |##########| Elapsed Time: 0:00:17 Time:  0:00:17


...done, 519873399 bytes.
23


100% (661005846 of 661005846) |##########| Elapsed Time: 0:01:23 Time:  0:01:23


...done, 661005846 bytes.
24


100% (588970092 of 588970092) |##########| Elapsed Time: 0:00:55 Time:  0:00:55


...done, 588970092 bytes.
25


100% (555920178 of 555920178) |##########| Elapsed Time: 0:00:52 Time:  0:00:52


...done, 555920178 bytes.
26


100% (501824629 of 501824629) |##########| Elapsed Time: 0:01:09 Time:  0:01:09


...done, 501824629 bytes.
27


100% (493401345 of 493401345) |##########| Elapsed Time: 0:00:18 Time:  0:00:18


...done, 493401345 bytes.
28


100% (476838637 of 476838637) |##########| Elapsed Time: 0:00:54 Time:  0:00:54


...done, 476838637 bytes.
29


100% (504273590 of 504273590) |##########| Elapsed Time: 0:00:47 Time:  0:00:47


...done, 504273590 bytes.
30


100% (537158203 of 537158203) |##########| Elapsed Time: 0:01:03 Time:  0:01:03


...done, 537158203 bytes.
31


100% (524165119 of 524165119) |##########| Elapsed Time: 0:00:55 Time:  0:00:55


...done, 524165119 bytes.
32


100% (604200131 of 604200131) |##########| Elapsed Time: 0:01:11 Time:  0:01:11


...done, 604200131 bytes.
33


100% (449703847 of 449703847) |##########| Elapsed Time: 0:00:42 Time:  0:00:42


...done, 449703847 bytes.
34


100% (545875287 of 545875287) |##########| Elapsed Time: 0:00:21 Time:  0:00:21


...done, 545875287 bytes.
35


100% (534396419 of 534396419) |##########| Elapsed Time: 0:01:23 Time:  0:01:23


...done, 534396419 bytes.
36


100% (595435274 of 595435274) |##########| Elapsed Time: 0:00:25 Time:  0:00:25


...done, 595435274 bytes.
37


HTTPError: HTTP Error 404: The specified blob does not exist.

## Code to Fix arrays 

Fix arrays when tiles are not properly allocated

In [ ]:
N =   #specify the number of tiles that were not properly allocated
#load the array
remaining = np.load('tile_name_tile_url_remaining_expanded.npy')
labeled = np.load("tile_name_tile_url_labeled.npy")
#Save the ammended arrays
np.save('tile_name_tile_url_labeled',labeled[:-N])
np.save('tile_name_tile_url_remaining_expanded',np.concatenate((labeled[-N:], remaining)))

Fix arrays when a tile cannot be found (HTTP Error 404: The specified blob does not exist.)

In [16]:
row_to_remove = 53

#dist.tile_name_tile_url_tiles_for_annotators[row_to_remove]

error_downloading = np.concatenate((np.load('tile_name_tile_url_error_downloading.npy'),
                                    dist.tile_name_tile_url_tiles_for_annotators[row_to_remove]))
np.save('tile_name_tile_url_error_downloading.npy', error_downloading)

#Remove the rows with errors
print(np.load('tile_name_tile_url_remaining_expanded.npy').shape)
np.save('tile_name_tile_url_remaining_expanded',np.delete(np.load(tiles_remaining), row_to_remove, axis = 0))
print(np.load('tile_name_tile_url_remaining_expanded.npy').shape)
row_to_remove = np.nan

(595, 2)
(594, 2)


In [ ]:
#recover the unlabeled images from the dist
## np.save('tile_name_tile_url_remaining_expanded',dist.tile_name_tile_url_unlabeled)
#dist.tile_name_tile_url_tiles_for_annotators

In [ ]:
#dist.track_tile_annotations(tiles_labeled)
#np.save('tile_name_tile_url_remaining_expanded', dist.tile_name_tile_url_remaining)
#np.save('tile_name_tile_url_labeled', dist.tile_name_tile_url_labeled)

## FIX Crop Script

In [ ]:
from PIL import Image

def crop(path, input, height, width, k, page, area):
    im = Image.open(input)
    imgwidth, imgheight = im.size
    for i in range(0,imgheight,height):
        for j in range(0,imgwidth,width):
            box = (j, i, j+width, i+height)
            a = im.crop(box)
            try:
                o = a.crop(area)
                o.save(os.path.join(path,"PNG","%s" % page,"IMG-%s.png" % k))
            except:
                pass
            k +=1

In [ ]:
dist.tile_names = os.listdir(dist.tiles_dir) #get a list of all of the tiles in tiles directory
for tile_name in dist.tile_names: #index over the tiles in the tiles_dir 
    file_name, ext = os.path.splitext(tile_name) # File name
    print(tile_name)

    print(count)

### Pil

In [ ]:
from PIL import Image 
 
#img = Image.open(r"C:\Users\rapiduser\Downloads\shutterstock_95235265.jpg") 
img = Image.open(r"C:\chips_allocation\Check\tiles\ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif")
img.save('test.jpeg')
img = Image.open(r"C:\chips_allocation\Check\tiles\ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.")

left = 0
top = 50
right = 510
bottom = 292
 
img_res = img.crop((left, top, right, bottom)) 
 
img_res.show() 

### Open CV

In [14]:
import cv2
import os
image = cv2.imread(r"C:/chips_allocation/Check/tiles/ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif") 
y=0
x=512*19
h=512
w=512
#cropped = image[x:w, y:h]
cropped = image[y:y+h, x:x+w]
print(image.shape)

# Filename
filename = 'test2.jpg'
cv2.imwrite(os.path.join('C:/chips_allocation/Check/tiles',filename),cropped)
cv2.imshow("Cropped", cropped)
cv2.waitKey(0)
# Using cv2.imwrite() method
# Saving the image
#cv2.imwrite(filename, img)


(12220, 9770, 3)


-1

In [ ]:
from PIL import Image
Image.MAX_IMAGE_PIXELS = 999999999999
import os
file_name = "ca_060cm_2018_38121_m_3812128_ne_10_060_20180714"
#tile = Image.open(os.path.join(dist.tiles_dir, tile_name)) #open image
with Image.open(r"C:\chips_allocation\Check\tiles\ca_060cm_2018_38121_m_3812128_ne_10_060_20180714") as tile:
    width, height = tile.size
    item_width = int(512)
    col = int(width/512)+1
    row = int(height/512)+1
    count = 1
    for j in range(0,col):
        for i in range(0,row):
            box = (i*item_width, j*item_width, (i+1)*item_width, (j+1)*item_width)
            chips_save_path = os.path.join(self.chips_dir, file_name+ '_'+ \
                                           str(count).zfill(6) + '.jpg') 
            # The index is a six-digit number like '000023'. 
            #chip = tile.crop(box)
            #chip = chip.save(chips_save_path)
            chip = tile.crop(box).save(chips_save_path)
            count += 1
    print(count)

In [4]:
import numpy as np
import cv2
import os

image = cv2.imread(r"C:/chips_allocation/Check/tiles/ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif") 
y=0
x=512*19
h=512
w=512
#cropped = image[x:w, y:h]
cropped = image[y:y+h, x:x+w]

print(image.shape)

# Filename
filename = 'test2.jpg'
cv2.imwrite(os.path.join('C:/chips_allocation/Check/tiles',filename),cropped)
cv2.imshow("Cropped", cropped)
cv2.waitKey(0)
# Using cv2.imwrite() method
# Saving the image
#cv2.imwrite(filename, img)

#image = cv2.imread('C:/chips_allocation/Check/tiles/test2.jpg') 
print(chip_img.width)
print(chip_img.height)

#At the end of the image
if chip_img.width != 512:
    black_width = 512 - chip_img.width
    black_height = 512
    black_image = np.zeros((black_height,black_width,3), np.uint8)
    print("Incorrect Width")
    print(black_img.width)
    print(black_img.height)
    chip_img = np.concatenate([chip_img, blank_image])
    
if chip_img.height != 512:
    black_height = 512 - chip_img.height
    black_width = 512
    black_image = np.zeros((black_height,black_width,3), np.uint8)
    print("Incorrect Height")
    print(black_img.width)
    print(black_img.height)
    chip_img = np.concatenate([chip_img, blank_image])

In [8]:
#blank_image = np.zeros((512,512-42,3), np.uint8)
cv2.imshow('3 Channel Window', 1)
print("image shape: ", blank_image.shape)
cv2.waitKey(0)
cv2.destroyAllWindows()

image shape:  (512, 470, 3)


In [35]:
from PIL import Image


#blackBox = Image.new(im.mode, (im.width, 50), '#000')
#im.paste(blackBox, (0, im.height - blackBox.height))
#for x in range(img.width):
#    for y in range(img.height - 40, img.height):
#        img.putpixel((x, y), (0, 0, 0))

#img.save('test2_bb.png')

42
512


In [40]:
from PIL import Image  
chip_img = Image.open('C:/chips_allocation/Check/tiles/test2.jpg')
print(chip_img.width)
print(chip_img.height)

if chip_img.width != 512:
    black_width = 512 - chip_img.width
    black_height = 512
    black_img = Image.new(mode = "RGB", size = (black_width, black_height))
    print("Incorrect Width")
    print(black_img.width)
    print(black_img.height)
    chip_img+black_img
    
if chip_img.height != 512:
    black_height = 512 - chip_img.height
    black_width = 512
    blackimg = Image.new(mode = "RGB", size = (black_width, black_height))
    print("Incorrect Height")
    print(black_img.width)
    print(black_img.height)

42
512
Incorrect Width
470
512


TypeError: unsupported operand type(s) for +: 'JpegImageFile' and 'Image'

In [ ]:
import numpy as np
import cv2

image = cv2.imread('download.jpg')
y=0
x=0
h=100
w=200
crop = image[y:y+h, x:x+w]
cv2.imshow('Image', crop)
cv2.waitKey(0) 

### Wand

In [7]:
from __future__ import print_function
from wand.image import Image

with Image(filename=r"C:/chips_allocation/Check/tiles/ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif") as img:
    print('format =', img.format)
    print('width =', img.width)
    print('height =', img.height)

format = TIFF
width = 306
height = 382
